File: clients/ios/Lextures/Core/LMS/WalletLogic.swift
import Foundation

/// Credentials wallet helpers (M12.2).
enum WalletLogic {
    static func walletEnabled(_ features: MobilePlatformFeatures) -> Bool {
        credentialsSectionEnabled(features)
            || ccrEnabled(features)
            || ceTranscriptEnabled(features)
            || officialTranscriptsEnabled(features)
    }

    static func credentialsSectionEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffCompletionCredentials
    }

    static func ccrEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffCoCurricularTranscript
    }

    static func ceTranscriptEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffCeuTracking
    }

    static func officialTranscriptsEnabled(_ features: MobilePlatformFeatures) -> Bool {
        features.ffTranscripts
    }

    static func cacheKeySummary() -> String { "wallet:summary" }
    static func cacheKeyCCR() -> String { "wallet:ccr" }
    static func cacheKeyCETranscript() -> String { "wallet:ce-transcript" }
    static func cacheKeyTranscriptRequests() -> String { "wallet:transcript-requests" }

    static func officialTranscriptWebURL() -> URL {
        AppConfiguration.webURL(path: "/transcripts")
    }

    static func dateLabel(iso: String) -> String {
        CredentialsLogic.issuedDateLabel(iso: iso)
    }

    static func achievementTypeLabel(_ type: String) -> String {
        type.replacingOccurrences(of: "_", with: " ").capitalized
    }

    static func transcriptStatusLabel(_ status: String) -> String {
        switch status {
        case "queued":
            return L.text("mobile.wallet.requestStatus.queued")
        case "submitted":
            return L.text("mobile.wallet.requestStatus.submitted")
        case "failed":
            return L.text("mobile.wallet.requestStatus.failed")
        default:
            return status
        }
    }

    static func deliveryTypeLabel(_ type: String) -> String {
        switch type.lowercased() {
        case "email":
            return L.text("mobile.wallet.delivery.email")
        case "mail":
            return L.text("mobile.wallet.delivery.mail")
        case "pickup":
            return L.text("mobile.wallet.delivery.pickup")
        default:
            return type
        }
    }

    static func ccrPdfPreviewTarget(documentId: String) -> FilePreviewTarget {
        FilePreviewTarget(
            courseCode: "_wallet",
            displayName: "ccr.pdf",
            mimeType: "application/pdf",
            byteSize: nil,
            source: .directPath(LMSAPI.ccrDownloadPath(documentId: documentId, format: "pdf"))
        )
    }

    static func ceTranscriptPdfPreviewTarget() -> FilePreviewTarget {
        FilePreviewTarget(
            courseCode: "_wallet",
            displayName: "ce-transcript.pdf",
            mimeType: "application/pdf",
            byteSize: nil,
            source: .directPath(LMSAPI.ceTranscriptPdfPath())
        )
    }
}

File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/WalletLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.navigation.MobilePlatformFeatures
import com.lextures.android.core.config.AppConfiguration

/** Credentials wallet helpers (M12.2). */
object WalletLogic {
    fun walletEnabled(features: MobilePlatformFeatures): Boolean =
        credentialsSectionEnabled(features) ||
            ccrEnabled(features) ||
            ceTranscriptEnabled(features) ||
            officialTranscriptsEnabled(features)

    fun credentialsSectionEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffCompletionCredentials

    fun ccrEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffCoCurricularTranscript

    fun ceTranscriptEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffCeuTracking

    fun officialTranscriptsEnabled(features: MobilePlatformFeatures): Boolean =
        features.ffTranscripts

    fun cacheKeyCcr(): String = "wallet:ccr"
    fun cacheKeyCeTranscript(): String = "wallet:ce-transcript"
    fun cacheKeyTranscriptRequests(): String = "wallet:transcript-requests"

    fun officialTranscriptWebUrl(): String = AppConfiguration.webUrl("/transcripts")

    fun dateLabel(iso: String): String = CredentialsLogic.issuedDateLabel(iso)

    fun achievementTypeLabel(type: String): String =
        type.replace('_', ' ').replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }

    fun transcriptStatusLabel(status: String): String = when (status) {
        "queued" -> "Queued"
        "submitted" -> "Submitted"
        "failed" -> "Failed"
        else -> status
    }

    fun deliveryTypeLabel(type: String): String = when (type.lowercase()) {
        "email" -> "Email"
        "mail" -> "Mail"
        "pickup" -> "Pickup"
        else -> type
    }

    fun ccrPdfPreviewTarget(documentId: String): FilePreviewTarget =
        FilePreviewTarget.submissionContentPath(
            courseCode = "wallet",
            contentPath = LmsApi.ccrDownloadPath(documentId, "pdf"),
            fileName = "ccr.pdf",
            mimeType = "application/pdf",
        )

    fun ceTranscriptPdfPreviewTarget(): FilePreviewTarget =
        FilePreviewTarget.submissionContentPath(
            courseCode = "wallet",
            contentPath = LmsApi.ceTranscriptPdfPath(),
            fileName = "ce-transcript.pdf",
            mimeType = "application/pdf",
        )
}
